CARDS-1851: Add Cancel button to all single-configuration admin screens

marta- opened the pull request:
**Includes:**
* [CARDS-1851](https://phenotips.atlassian.net/browse/CARDS-1851) - _Add Cancel button to all single-configuration admin screens_
* Refactoring of all single-config admin screens with the introduction of a new component `AdminConfigScreen` that handles loading, saving, displaying buttons
* `fetchWithReLogin` is now used in these screens for loading and saving the configuration instead of simple `fetch`
* The `Save` button is now `disabled` when there are no pending changes instead of having the label `Saved`

**Testing:**
* Start in `proms` mode
* Navigate to `Administration`
* Test the entire functionality (loading the config, changing and saving the config, using the config) of these admin sections:
  * Quick search
  * Downtime warning
  * Patient Portal Welcome Message
  * Patient Portal Terms of Use

**Reviewing:**
* [Ignoring whitespace changes](https://github.com/data-team-uhn/cards/pull/1110/files?w=1) might help see better what parts of the affected admin section components have been moved to `AdminConfigScreen`.

marta- committed:
CARDS-1851: Add Cancel button to all single-configuration admin screens
To avoid duplicate code, all single-config admin screens have been refactored, with the introduction of a new component `AdminConfigScreen` that handles loading, saving, displaying buttons.
Other changes:
* `fetchWithReLogin` is now used in these screens for loading and saving the configuration instead of simple `fetch`
* The Save button is now disabled when there are no pending changes instead of having the label "Saved"

marta- commented on modules/homepage/src/main/frontend/src/adminDashboard/AdminConfigScreen.jsx:
```diff
@@ -0,0 +1,139 @@
+}));
+
+function AdminConfigScreen(props) {
+  const { title, configPath, onConfigFetched, hasChanges, configError, buildConfigData, onConfigSaved, children } = props;
```
@sdumitriu Should `buildConfigData` be named `onBeforeSubmit` instead?

sdumitriu replied:
`on*` usually means that it's just a push notification without an expected return, so I wouldn't call it `onBeforeSubmit`. `buildConfigData` sounds good enough.

sdumitriu commented on modules/downtime-warning-banner/src/main/frontend/src/DowntimeWarningConfiguration.jsx:
```diff
@@ -16,130 +16,79 @@
   const [ fromDate, setFromDate ] = useState();
   const [ toDate, setToDate ] = useState();
-  const [ isInvalidDateRange, setIsInvalidDateRange ] = useState(false);
+  const [ isDateRangeInvalid, setIsDateRangeInvalid ] = useState(false);
```
It is a convention that methods start with verbs. Since we're renaming this, can it be `invalidDateRange` or `dateRangeIsInvalid`?

marta- replied:
Renamed to `dateRangeIsInvalid` in 0812799.

sdumitriu commented on modules/downtime-warning-banner/src/main/frontend/src/DowntimeWarningConfiguration.jsx:
```diff
@@ -16,130 +16,79 @@
-  const [ path, setPath ] = useState();
-
-  // The the configuration values specified by the Administration
+  // The the configuration values
```
```suggestion
  // The configuration values
```

marta- replied:
Typo fixed in 0812799.

sdumitriu commented on modules/downtime-warning-banner/src/main/frontend/src/DowntimeWarningConfiguration.jsx:
```diff
@@ -16,130 +16,79 @@
-  const [ isSaved, setIsSaved ] = useState(false);
-  const [ fetched, setFetched ] = useState(false);
+  // Tracking unsaved changes
+  const [ hasChanges, setHasChanges ] = useState(true);
```
Why is it `true` initially?

marta- replied:
I think it was part of quickly converting isSaved (false) to hasChanges. Will change.

marta- replied:
Initial state left `undefined` since 0812799.

sashaandjic commented:
There is one inconsistency. On the `Quick search` page, `SAVE` button is enable once changes are made. The other 3 do not have this feature.
I personally prefer the behavior in the `Quick search`.

sdumitriu commented on modules/homepage/src/main/frontend/src/themePage/QuickSearchConfiguration.jsx:
```diff
@@ -28,77 +27,36 @@ import {
-      });
   }
 
   let onSourceTypeChange  = (checked, resourceName) => {
```
Probably not new in this PR, but I noticed the following bug: toggling items on or off requires clicking twice.

marta- replied:
I wonder if it's the `event.preventDefault` I removed. Will check.

marta- replied:
It was the `event.preventDefault` that I **didn't** remove.

marta- replied:
Fixed by removing `event.preventDefault()` (next line) in 0812799.

sdumitriu commented on modules/homepage/src/main/frontend/src/themePage/QuickSearchConfiguration.jsx:
```diff
@@ -108,13 +66,19 @@ function QuickSearchConfiguration(props) {
+    <AdminConfigScreen
+        title="Quick Search Settings"
+        configPath="/apps/cards/config/QuickSearch"
+        onConfigFetched={getQuickSearchSettings}
```
I think all the `get*` methods now have a misleading name and should be changed to `onConfigFetched`.

marta- replied:
I find that the code is more readable if the handler function says what it's doing, it reads more like natural language. `onConfigFetched={getSettings}` means _"When the configuration is fetched, get the settings [from it]_" (rather than `onConfigFetched={onConfigFetched}` -> "_When the configuration is fetched, do what you do when the configuration is fetched_"). Would it help if we rename `getQuickSearchSettings` to `readQuickSearchSettings`?

sdumitriu replied:
Yes, `read*` is better. Also look at the callbacks in the other screens.

marta- replied:
All will be renamed.

marta- replied:
Renamed all `onConfigFetched` handlers from `get*` to `read*`  in 0812799.

marta- committed:
CARDS-1851: Add Cancel button to all single-configuration admin screens
Addressing review comments:
* Fixed typo in comment
* Renamed boolean variable to avoid it being confused for a function
* Renamed handler functions for clarity
* Bonus: fixed "double-click to change" bug in quick search config checkboxes

marta- committed:
CARDS-1851: Add Cancel button to all single-configuration admin screens
Addressing review comments:
* Turned the cancel button into a Reset button that restores the settings first loaded on the page
* When Reset is clicked, a confirmation dialog appears informing the user about the consequesces of resetting and presenting possible next steps (reset, drop unsaved changes, cancel)

marta- commented:
> @marta On a second though, `CLOSE` should behave same way as `SAVE` ---> action, disable button and no closing. Enable button again once changes are made.

Re-implemented as a Reset button in c148983:
* Restores the settings first loaded on the page
* When clicked, a confirmation dialog appears informing the user about the consequences of resetting and presenting possible next steps (reset, drop unsaved changes, cancel)

sashaandjic commented:
After clicking on `YES, RESET` the dialog should close.
![image](https://user-images.githubusercontent.com/23243244/177915000-e4767303-8a53-476a-bfb3-2d9271b121a4.png)

sashaandjic commented:
Here is a really weird thing that sometimes happens. In the attached video, everything seems to work as expected until the very end when action of clicking on the `YES, RESET` button, would not disable `RESET TO INITIAL SETTINGS` button. The button gets disabled after second click on `YES, RESET`.

https://user-images.githubusercontent.com/23243244/177916994-7ef69717-570b-40ec-88df-d095dbefd6bf.mp4

marta- committed:
CARDS-1851: Add Cancel button to all single-configuration admin screens

Adressing review comments: add omitted line to close the reset confirmation dialog after clicking yes, reset

marta- commented:
> After clicking on `YES, RESET` button, the `Confirm configuration reset` dialog should close.

Added forgotten line to close the dialog in 895e206.

Probably outside of the scope, I would make `SAVE` and `RESET to INITIAL SETTINGS` buttons disabled when the page initially open.

I had them disabled for a while and thought it looked broken with all bittons disabled.

sashaandjic reviewed (approved):
tested

marta- committed:
CARDS-1851: Add Cancel button to all single-configuration admin screens
Addressing review comments: document the AdminConfigScreen component

marta- commented:
@sdumitriu Updated the documentation for `AdminConfigScreen` in 7e6f8bf.

@sashaandjic The latest changes are comments and proptypes and don't really need retesting.

sdumitriu reviewed (approved):
Idea for the future: for simple configs, let the `AdminConfigScreen` itself display all the properties as a plain input. Would need a way to figure out where to store the labels for the inputs, if a simple camel-case-to-sentence-case doesn't work.
